cluster_tools/cluster_tools/schedulers/slurm.py

```diff
@@ -411,69 +413,110 @@ def matches_states(slurm_states: List[str]) -> bool:
     def investigate_failed_job(
         self, job_id_with_index: str
     ) -> Optional[Tuple[str, Type[RemoteException]]]:
-        # We call `seff job_id` which should return some output including a line,
-        # such as: "Memory Efficiency: 25019.18% of 1.00 GB"
+        # This function tries to find the reason for a failed job by first checking whether
+        # the job run time exceeded the specified time limit. If that is not the case, it
+        # checks whether the job used too much RAM. As a last resort, it checks the exit code
+        # of the job: If the job was killed with signal 9, it's very likely due to some
+        # resource limit.
+
+        def parse_key_value_pairs(
+            pair_delimiter: str, key_value_delimiter: str
+        ) -> Dict[str, str]:
+            properties = {}
+            for key_value_pair in stdout.split(pair_delimiter):
```

Comment:
I stumbled a bit here, because this function reads `stdout` which is defined below said function. how about passing it as an input parameter to the function?

Reply:
Oh, yes that was by accident! I'll pass it in as a parameter.